server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -312,62 +311,69 @@ private SoftDeletesPolicy newSoftDeletesPolicy() throws IOException {
-                indexReader.incRef(); // steal the reader - getSearcher will decrement if it fails
-                current = SearcherManager.getSearcher(searcherFactory, indexReader, null);
+                acquire.incRef(); // steal the reader -  notifyListener will decrement if it fails
+                notifyListener(acquire, null);
+                current = acquire;
             } finally {
-                internalSearcherManager.release(acquire);
+                internalReaderManager.release(acquire);
             }
-            this.searcherFactory = searcherFactory;
-            this.internalSearcherManager = internalSearcherManager;
         }
 
         @Override
-        protected IndexSearcher refreshIfNeeded(IndexSearcher referenceToRefresh) throws IOException {
+        protected ElasticsearchDirectoryReader refreshIfNeeded(ElasticsearchDirectoryReader referenceToRefresh) throws IOException {
             // we simply run a blocking refresh on the internal reference manager and then steal it's reader
             // it's a save operation since we acquire the reader which incs it's reference but then down the road
             // steal it by calling incRef on the "stolen" reader
-            internalSearcherManager.maybeRefreshBlocking();
-            IndexSearcher acquire = internalSearcherManager.acquire();
+            internalReaderManager.maybeRefreshBlocking();
+            ElasticsearchDirectoryReader acquire = internalReaderManager.acquire();
             try {
-                final IndexReader previousReader = referenceToRefresh.getIndexReader();
-                assert previousReader instanceof ElasticsearchDirectoryReader:
-                    "searcher's IndexReader should be an ElasticsearchDirectoryReader, but got " + previousReader;
-
-                final IndexReader newReader = acquire.getIndexReader();
-                if (newReader == previousReader) {
+                if (acquire == referenceToRefresh) {
                     // nothing has changed - both ref managers share the same instance so we can use reference equality
                     return null;
                 } else {
-                    newReader.incRef(); // steal the reader - getSearcher will decrement if it fails
-                    return SearcherManager.getSearcher(searcherFactory, newReader, previousReader);
+                    acquire.incRef(); // steal the reader - notifyListener will decrement if it fails
+                    notifyListener(acquire, referenceToRefresh);
+                    return acquire;
                 }
             } finally {
-                internalSearcherManager.release(acquire);
+                internalReaderManager.release(acquire);
+            }
+        }
+
+        private void notifyListener(ElasticsearchDirectoryReader reader, ElasticsearchDirectoryReader previousReader) throws IOException {
```

Comment:
can we move the reference stealing into this method too?

Reply:
then we can also do this `return incrementAndNotify(acquire, referenceToRefresh)`